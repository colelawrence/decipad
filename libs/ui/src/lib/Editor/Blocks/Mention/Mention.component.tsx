import { TagLabel, Text } from '@chakra-ui/react';

import { RenderElementProps } from 'slate-react';
import { TagStyles } from './Mention.styles';

export const Mention = ({
  attributes,
  children,
  element,
}: RenderElementProps): JSX.Element => {
  return (
    <Text as="span" {...attributes}>
      <TagStyles contentEditable={false}>
        <TagLabel>@{element.user}</TagLabel>
      </TagStyles>
      {children}
    </Text>
  );
};
