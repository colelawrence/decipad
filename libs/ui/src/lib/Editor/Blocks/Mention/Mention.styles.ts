import { Tag } from '@chakra-ui/react';
import { chakra } from '@chakra-ui/system';

export const TagStyles = chakra(Tag, {
  baseStyle: {
    size: 'md',
    colorScheme: 'blue',
    borderRadius: 'full',
  },
});
