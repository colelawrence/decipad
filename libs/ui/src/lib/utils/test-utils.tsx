import { ChakraProvider } from '@chakra-ui/react';
import { render, RenderOptions } from '@testing-library/react';
import React from 'react';
import { theme } from '../theme';

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <ChakraProvider resetCSS theme={theme}>
      {children}
    </ChakraProvider>
  );
};

const customRender = (
  ui: JSX.Element,
  options?: Omit<RenderOptions, 'queries'>
) => render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
