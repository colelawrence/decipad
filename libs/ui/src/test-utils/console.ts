export const mockConsoleWarn = () => {
  const originalConsoleWarn = console.warn;
  const mockConsoleWarn: jest.MockedFunction<typeof console.warn> = jest.fn();

  beforeEach(() => {
    console.warn = jest.fn();
    mockConsoleWarn.mockClear();
  });
  afterEach(() => {
    console.warn = originalConsoleWarn;
  });

  return mockConsoleWarn;
};
