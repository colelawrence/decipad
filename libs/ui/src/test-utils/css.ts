import cssVarsPonyfill from 'css-vars-ponyfill';

export const findParentWithStyle = <P extends keyof CSSStyleDeclaration>(
  element: Element | null,
  propertyName: P
):
  | (Pick<CSSStyleDeclaration, P> & {
      element: Element;
      styles: CSSStyleDeclaration;
    })
  | null => {
  if (element === null) {
    return null;
  }
  const styles = getComputedStyle(element);
  if (!styles[propertyName]) {
    return findParentWithStyle(element.parentElement, propertyName);
  }
  return {
    [propertyName]: styles[propertyName],
    element,
    styles,
    // cannot type dynamic property key based on type parameter
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } as any;
};

export const applyCssVars = (): Promise<() => void> => {
  return new Promise((resolve, reject) => {
    const cleanup = () =>
      [...document.head.querySelectorAll('[data-cssvars="out"]')].forEach(
        (element) => element.remove()
      );
    cssVarsPonyfill({
      onError: reject,
      onComplete: () => resolve(cleanup),
    });
  });
};

export const applyPrefersColorScheme = async (
  colorScheme: 'dark' | 'light'
): Promise<() => void> => {
  // JSDOM ignores CSSOM modifications, so build a CSS string and create and append a style element instead
  const styleText = [...document.styleSheets]
    .flatMap((styleSheet) => [...styleSheet.cssRules])
    .flatMap((rule) => {
      if (rule.type === CSSRule.MEDIA_RULE) {
        const mediaRule = rule as CSSMediaRule;
        for (
          let mediaIndex = 0;
          mediaIndex < mediaRule.media.length;
          mediaIndex += 1
        ) {
          if (
            mediaRule.media[mediaIndex].replace(/\s/g, '') ===
            `(prefers-color-scheme:${colorScheme})`
          ) {
            return [...mediaRule.cssRules];
          }
        }
      }
      return [];
    })
    .flatMap((rule) => rule.cssText)
    .join(' ');

  const styleElement = document.createElement('style');
  styleElement.innerHTML = styleText;
  document.head.appendChild(styleElement);

  const cleanupCssVars = await applyCssVars();

  return () => {
    cleanupCssVars();
    styleElement.remove();
  };
};
