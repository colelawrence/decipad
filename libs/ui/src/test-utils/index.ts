export * from './menu';
export * from './language';
