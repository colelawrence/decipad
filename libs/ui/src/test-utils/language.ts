import {
  Result,
  runCode as run,
  SerializedTypeKind,
  serializeResult,
} from '@decipad/computer';

export async function runCode<T extends SerializedTypeKind>(
  code: string
): Promise<Result.Result<T>> {
  const result = await run(code);
  return serializeResult<T>(result.type, result.value);
}
