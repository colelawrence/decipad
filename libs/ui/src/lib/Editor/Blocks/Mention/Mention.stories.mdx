import { Meta, Story, Canvas, Props } from '@storybook/addon-docs/blocks';
import { Mention } from './Mention.component';

<Meta
  title="Editor/Blocks/Mention"
  parameters={{
    docs: {
      source: {
        code: `<Mention
  element={{
    children: [{ text: 'Hello World' }],
    user: 'johndoe'
  }}
  attributes={{ 'data-slate-node': 'element', ref: null }}
>
  Hello World
</Mention>`,
      },
    },
  }}
  argTypes={{
    children: {
      defaultValue: 'Hello World',
      description: 'The text content of the component',
      type: 'string',
    },
    element: {
      defaultValue: { user: 'johndoe' },
      description: 'The user mentioned',
      type: {
        user: 'string',
      },
    },
  }}
  component={Mention}
/>

# Mention

The mention component is a slate inline block that is used to mention other users throughout the editor.

export const Template = (args) => <Mention {...args} />;

<Canvas>
  <Story name="Mention">{Template.bind({})}</Story>
</Canvas>

<Props story="Mention" />
