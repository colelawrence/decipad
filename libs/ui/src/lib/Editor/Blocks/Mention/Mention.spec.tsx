import React from 'react';
import { render } from 'test-utils';
import { Mention } from './Mention.component';

describe('Mention Inline Block', () => {
  it('renders', () => {
    const { container } = render(
      <Mention
        element={{
          children: [{ text: 'Hello World' }],
          user: 'johndoe',
        }}
        attributes={{ 'data-slate-node': 'element', ref: null }}
      >
        Next text
      </Mention>
    );

    expect(container).toBeDefined();
  });

  it('matches snapshot', () => {
    const { container } = render(
      <Mention
        element={{
          children: [{ text: 'Hello World' }],
          user: 'johndoe',
        }}
        attributes={{ 'data-slate-node': 'element', ref: null }}
      >
        Next text
      </Mention>
    );

    expect(container).toMatchSnapshot();
  });

  it('renders the user tag and continues to the next text', () => {
    const { getByText } = render(
      <Mention
        element={{
          children: [{ text: 'Hello World' }],
          user: 'johndoe',
        }}
        attributes={{ 'data-slate-node': 'element', ref: null }}
      >
        Next text
      </Mention>
    );

    getByText('@johndoe');
    getByText('Next text');
  });
});
