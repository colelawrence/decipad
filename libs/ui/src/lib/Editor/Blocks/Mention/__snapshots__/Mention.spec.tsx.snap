// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mention Inline Block matches snapshot 1`] = `
.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: top;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: 100%;
  font-weight: var(--chakra-fontWeights-medium);
  line-height: 1.2;
  outline: 2px solid transparent;
  outline-offset: 2px;
  min-height: 1.5rem;
  min-width: 1.5rem;
  font-size: var(--chakra-fontSizes-sm);
  border-radius: var(--chakra-radii-md);
  -webkit-padding-start: var(--chakra-space-2);
  padding-inline-start: var(--chakra-space-2);
  -webkit-padding-end: var(--chakra-space-2);
  padding-inline-end: var(--chakra-space-2);
  background: var(--chakra-colors-gray-100);
  color: var(--chakra-colors-gray-800);
  size: md;
  -webkit-print-color-scheme: blue;
  color-scheme: blue;
  border-radius: var(--chakra-radii-full);
}

.emotion-1:focus,
.emotion-1[data-focus] {
  box-shadow: var(--chakra-shadows-outline);
}

.emotion-2 {
  line-height: 1.2;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<div>
  <span
    class="chakra-text emotion-0"
    data-slate-node="element"
  >
    <span
      class="emotion-1"
      contenteditable="false"
    >
      <span
        class="emotion-2"
      >
        @
        johndoe
      </span>
    </span>
    Next text
  </span>
</div>
`;
